### PR TITLE
[Popover] more ways to popover: two children or content/target props

### PR DESCRIPTION
**Pull request by giladgray**
#### Changes proposed in this pull request:

__BLUF__: new `target` prop and support for a second child instead of `content` prop. enables prop-only or children-only usage.

When creating a popover, you must specify both its _content_ and its _target_.
This can be done a few ways:

1. Provide both the `content` and `target` props, which accept a string or a JSX element.
    Omitting the `target` prop will produce an error.
    ```tsx
    <Popover content={<Content />} target={<Button text="Open" />} />
    ```

1. Provide one or two `children`. Omitting a `target` element will produce an error.
    ```tsx
    <Popover>
        <Button text="Open" />
        <Content />
    </Popover>
    ```

1. It is possible to mix the two: provide the `content` prop and one React child as the target.
    (Using the `target` prop with `children` is not supported and will produce a warning.)
    ```tsx
    <Popover content={<Content />}>
        <Button text="Open" />
    </Popover>
    ```

#### Rationale

1. I've always found `content` usage clunky... it often requires defining a local variable to write multiple lines of JSX. I like the way [react-tether does it](https://github.com/souporserious/react-tether#example-usage) with one or two children. Note that in react-tether, omitting the content element is equal to our `isOpen={false}`; In Blueprint, omitting the content will force-disable the Popover per #1094.

1. That said, sometimes it's nice to use props for everything _instead_ of children, so why not support that use case too! Hence the new `target` prop. **This also keeps this change backwards compatible!**

1. TypeScript does not enforce anything about `children` so I tried to design an API that does not rely on children to _look_ a certain way; it merely asserts the _number_ of children. For this reason, Popover supports string children too (which it wraps in a `span` internally). Also, validation is mostly warnings--the only unrecoverable error state is a missing target.

#### Reviewers should focus on:

Do you like this?

### CODE REVIEW

**Comment by adidahiya** [comment]
> TypeScript does not enforce anything about children so I tried to design an API that does not rely on children to look a certain way; it merely asserts the number of children. For this reason, Popover supports string children too (which it wraps in a span internally). Also, validation is mostly warnings--the only unrecoverable error state is a missing target.

big +1 for this. React children are hard to typedef.


```ts
<Popover>
    <Button text="Open" />
    <Content />
</Popover>
```

what is usage option 2 for? stylistic preference?

**Comment by cmslewis** [comment]
This is really cool! Looks reasonable, though Option 2 feels sneaky, precedented as it might be. Will let the discussion continue a little more before approving.

**Comment by giladgray** [comment]
@adidahiya option two supports inlining content in one big JSX block, rather than having to define a local variable for complex content.

```jsx
<Popover>
    <Button />
    <div>
        <h5>Popover title</h5>
        <a bunch of content elements ... />
        <and a button why not />
    </div>
</Popover>
```

**Comment by adidahiya** [comment]
ok I'm sold. I do see this pattern _a lot_.

```ts
const content = (
    ...
);

<Popover content={content} ...>
    <Button />
</Popover>
```

File: packages/core/src/common/errors.ts [hunk 1]
```diff
@@ -38,9 +38,17 @@ export const NUMERIC_INPUT_STEP_SIZE_NON_POSITIVE =
...
     `${ns} <NumericInput> requires stepSize to be defined.`;
 
-export const POPOVER_ONE_CHILD = `${ns} <Popover> requires exactly one target element`;
+export const POPOVER_REQUIRES_TARGET =
+    `${ns} <Popover> requires target prop or at least one child element.`;
 export const POPOVER_MODAL_INTERACTION =
     `${ns} <Popover isModal={true}> requires interactionKind={PopoverInteractionKind.CLICK}.`;
+export const POPOVER_WARN_TOO_MANY_CHILDREN =
+    `${ns} <Popover> supports one or two children; additional children are ignored.`
+    + ` First child is the target, second child is the content. You may instead supply these two as props.`;
```
Comment (cmslewis) [inline packages/core/src/common/errors.ts]: nit: Don't need back-ticks here.

File: packages/core/src/components/popover/popover.tsx [hunk 2]
```diff
@@ -437,13 +449,23 @@ export class Popover extends AbstractComponent<IPopoverProps, IPopoverState> {
...
                     <div className={Classes.POPOVER_CONTENT}>
-                        {this.props.content}
+                        {content}
                     </div>
                 </div>
             </div>
         );
     }
 
+    private understandChildren() {
```
Comment (tgreenwatts) [inline packages/core/src/components/popover/popover.tsx]: The Kids Are Alright

**Comment by noah79** [comment]
A think I ran into today is lack of ability to set the target component to something higher in the render tree.

My issue is that I have a component which has multiple popovers, however I want to position one of the popovers relative to the component as a whole.

Popover for row:
image

Nested tooltip:
image

I've implemented this currently by manually showing the popover via isOpen and setting the state when the user mouses in and out of the fluid label element (the triggering element).

I tried setting tetherOptions={{target: this.node}} but it's just ignored...

Some code:
```
return <sem.Menu.Item
			{...props}
			className={classNames(css.root, className, commonCss.menuItem, 'show-on-hover')}>
			<bp.Popover
				position={bp.Position.RIGHT}
				autoFocus={false}
				isOpen={showTooltip}
				popoverClassName={commonCss.sidebarPopover}
				className='fluid'
				interactionKind={bp.PopoverInteractionKind.HOVER}
				content={<QueryResultCard queryResult={qr} isTooltip/>}>
	<div className={commonCss.row}>
					<div className="fluid" onMouseEnter={() => this.showTooltip = true} onMouseLeave={() => this.showTooltip = false}>
						{isRenaming
							? <bp.EditableText className={commonCss.editableName}
							                   defaultValue={qr.name}
							                   selectAllOnFocus={true}
							                   ref={this.onEditRef}
							                   onCancel={qr.cancelRename}
							                   onConfirm={qr.confirmRename}/>
							: <Link className={commonCss.name} to={qr.clientUrl}>{name}{query && query.queryResultId == qr.id && "*"}</Link>
						}
					</div>

					<div className={classNames(commonCss.rightButtonGroup, "pt-button-group pt-minimal")}>
						{qr.isLoaded && <AvailableQueryViewDropdown position={bp.Position.BOTTOM_LEFT} queryResult={qr}/>}

						<bp.Tooltip position={bp.Position.BOTTOM} content={`Delete '${name}'`}>
							<bp.Button className="pt-minimal" iconName="small-cross" onClick={async () => await qr.delete()}/>
						</bp.Tooltip>

						{/*<sem.Label><sem.Icon name="database"/>{qr.simulation ? qr.simulation.name: <i>Deleted</i>}</sem.Label>*/}

					</div>
				</div>
			</bp.Popover>
		</sem.Menu.Item>```

**Comment by adidahiya** [comment]
@noah79 

> I've implemented this currently by manually showing the popover via isOpen and setting the state when the user mouses in and out of the fluid label element (the triggering element).

This is the recommended approach 👍 

I don't think Popover's uncontrolled mode implementation should become _even more complex_ to support this use case if controlled usage works well.

**Comment by giladgray** [comment]
@noah79 verbal 👍  to @adidahiya's comment above. controlled mode exists for those times when uncontrolled mode doesn't cut it (and don't forget about `onInteraction`).

**Pull request merged**